src/ir/access_path_root.h

```diff
@@ -0,0 +1,137 @@
+#ifndef SRC_IR_ACCESS_PATH_ROOT_H_
+#define SRC_IR_ACCESS_PATH_ROOT_H_
+
+#include "absl/types/variant.h"
+#include "absl/strings/str_join.h"
+#include "src/common/logging/logging.h"
+#include "third_party/arcs/proto/manifest.pb.h"
+
+// The classes in this file describe the root of an AccessPath. At the moment,
+// we have only two types of roots: a HandleConnectionSpecAccessPathRoot and
+// a HandleConnectionAccessPathRoot, describing a HandleConnectionSpec in a
+// ParticleSpec and a Handle in a Particle, respectively.
+namespace raksha::ir {
+
+// A HandleConnectionSpecAccessPathRoot describes the root as a
+// HandleConnectionSpec in a ParticleSpec. We consider this uninstantiated as
+// ParticleSpecs are abstract Particles which may be instantiated in multiple
+// locations.
+class HandleConnectionSpecAccessPathRoot {
+ public:
+  static HandleConnectionSpecAccessPathRoot CreateFromProto(
+      const arcs::AccessPathProto_HandleRoot &handle_root_proto) {
+    std::string particle_spec_name = handle_root_proto.particle_spec();
+    CHECK(!particle_spec_name.empty())
+      << "Expected a HandleRoot message to have a non-empty particle_spec.";
+    std::string handle_connection_spec_name =
+        handle_root_proto.handle_connection();
+    CHECK(!handle_connection_spec_name.empty())
+      << "Expected a HandleRoot message to have a non-empty handle_connection.";
+    return HandleConnectionSpecAccessPathRoot(
+        particle_spec_name, handle_connection_spec_name);
+  }
+
+  explicit HandleConnectionSpecAccessPathRoot(
```

Comment:
`explicit` is redundant, but OK if you prefer to leave it?

Reply:
Yeah, I try to keep in the habit of using `explicit` unless I want an implicit constructor, even in the cases where it's not necessary. Mostly so that I'm safe if I drop a parameter and forget to update the qualifier.